api/envoy/config/filter/http/transformation/v2/transformation_filter.proto

```diff
@@ -171,6 +171,12 @@ message Extraction {
   // If your regex contains capturing groups, use this field to determine which
   // group should be selected.
   uint32 subgroup = 3;
+
+  // The string to replace the matched portion of the source with
+  google.protobuf.StringValue replacement_text = 5;
+
+  // If set to true, all matches of the regex in the source will be replaced by the replacement_text.
```

Comment:
Could you explain what it does if it's false? Does it only replace the first instance then?

Reply:
That is essentially correct. We call this out in a couple of places in the docs, but by default an extractor's `regex` field needs to match the entire input (if you scroll down a bit [here](https://docs.solo.io/gloo-edge/1.7.23/guides/traffic_management/request_processing/transformations/#extractors) there's a comment speling this out)

```yaml
    # The regex must match the ENTIRE body, else the extraction will not work
    # The [\s\S]* allows us to match any character, making the regex match the entire source.
    regex: '[\s\S]*id: ([0-9]{4})[\s\S]*'
```

**so if `replace_all` is false and `replacement_text` is set, we replace the selected subgroup with the `replacement_text`** (this is the case you're calling out)

if `replace_all` is false and `replacement_text` is not set, we just use the pre-existing extractor functionality (i.e., extract the value of the selected subgroup)

if `replace_all` is true, we relax the restriction that the regex needs to match the entire source (I have to say, I don't love that restrictions around the regex change when this field is set). In this case, the regex should only partially match the input value. All matches against the input value will have their content replaced by `replacement_text`

note that if `replace_all` is true, `replacement_text` must be set 
 - I am going to validate that I test against this

Reply:
I updated the API to include a mode selection which should hopefully clarify some of these issues here: https://github.com/solo-io/envoy-gloo/pull/301/commits/3554ba1e82997a59633c4a17a7bc4011f1ac4ac0

```yaml
  enum Mode {
    // Default mode. Extract the value of the subgroup-th capturing group.
    EXTRACT = 0; 
    // Replace the value of the subgroup-th capturing group with the replacement_text.
    // Note: replacement_text must be set for this mode.
    SINGLE_REPLACE = 1;
    // Replace all matches of the regex in the source with the replacement_text.
    // Note: replacement_text must be set for this mode.
    // Note: subgroup is ignored for this mode. configuration will fail if subgroup is set.
    // Note: restrictions on the regex are different for this mode. See
    REPLACE_ALL = 2;
  }
```